text/0000-fragment-specifiers-for-generic-arguments.md

```diff
@@ -0,0 +1,155 @@
+  [`GenericParams`](https://doc.rust-lang.org/reference/items/generics.html)
+  grammar item.
+* `:generic_param` matches any of a lifetime, an identifier, or `const` followed
+  by an identifier.
+* `:generic_bound` matches the
+  [`TypeParamBounds`](https://doc.rust-lang.org/reference/trait-bounds.html)
+  (can be the bounds on a type parameter) or
+  [`LifetimeBounds`](https://doc.rust-lang.org/reference/trait-bounds.html) (can
+  be the bounds on a lifetime parameter) grammar items, or a type (can be the
+  bounds on a const parameter).
+* `:generic_default` matches a type (can be the default for a type parameter) or
+  anything that can be default for a const parameter (a block, an identifier, or
+  a literal).
+
+All of these can potentially pick up on multiple tokens, so the result of any of
+these parses is undestructible in the declarative macro.
+
+Following behavior:
+* `:generic` can be followed by anything, as it unambiguously ends when the
+  closing `>` appears.
+* `:generic_param` is similarly bounded and so anything can follow it, as well.
+* `:generic_bound` can be followed by anything that follows `:path` and `:ty`,
+  as it contains some repetition of lifetimes and paths separated by `+`, or a
+  type, and `+` is already illegal following a path or type.
+* `:generic_default` can be followed by anything that follows `:ty`, since the
+  other options all have an unambiguous end.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This provides more features which will need to be supported going forward. This
+also provides another features which "macros 2.0" will need to implement for
+parity with existing declarative macros.
+
+As far as I can tell, this additional cost to implementing and maintaining extra
+code is the only drawback associated with this feature.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We can do nothing, which provides no additional features and avoids the time and
+effort cost of implementing and maintaining this feature.
+
+# Prior art
+[prior-art]: #prior-art
+
+I'm not personally aware of any other languages that have similar declarative
+macros to Rust with an equivalent to fragment specifiers, nor any prior effort
+to add fragment specifiers covering this usage into Rust, so I don't know of any
+prior art on this topic.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Are these the best fragment specifiers to use for parsing macros? I'd like
+opinions from other people who write macros that would want something like this
+about if breaking up generics into some other form might be more useful. I think
+this set of fragment specifiers is the best for my use cases, but other people
+might be interested in macros that parse differently and they might want other
+things instead.
+
+Also, should `:generic_bound` include the preceding `:` in the match (e.g. `: 'a
++ SomeTrait` in the example above)? And likewise with the `=` before
```

Comment:
don't word-wrap in the middle of a \` code span, it doesn't get rendered properly

Reply:
Oops, I guess I need a better editor config. Fixed in d77ac900